src/modules/utils/GEN_CSV_READER_fbt.cpp

```diff
@@ -0,0 +1,189 @@
+    } else {
+      closeCSVFile();
+    }
+    sendOutputEvent(scmEventINITOID, paECET);
+  } else if(scmEventREQID == paEIID) {
+    QO() = QI();
+    if(QI()) {
+      readCSVFileLine();
+    }
+    sendOutputEvent(scmEventCNFID, paECET);
+  }
+}
+
+GEN_CSV_READER::GEN_CSV_READER(const CStringDictionary::TStringId paInstanceNameId, forte::core::CFBContainer &paContainer) :
+    CGenFunctionBlock<CFunctionBlock>(paContainer, paInstanceNameId), mDataOutputNames(nullptr), mDataOutputTypeIds(nullptr) {
+}
+
+GEN_CSV_READER::~GEN_CSV_READER(){
+  delete[] mDataOutputNames;
+  delete[] mDataOutputTypeIds;
+  closeCSVFile();
+}
+
+void GEN_CSV_READER::readInputData(TEventID paEI) {
+  switch(paEI) {
+    case scmEventINITID: {
+      readData(0, *mDIs[0], mDIConns[0]);
+      readData(1, *mDIs[1], mDIConns[1]);
+      break;
+    }
+    case scmEventREQID: {
+      readData(0, *mDIs[0], mDIConns[0]);
+      break;
+    }
+    default:
+      break;
+  }
+}
+
+void GEN_CSV_READER::writeOutputData(TEventID paEO) {
+  switch(paEO) {
+    case scmEventINITOID: {
+      writeData(0, *mDOs[0], mDOConns[0]);
+      writeData(1, *mDOs[1], mDOConns[1]);
+      break;
+    }
+    case scmEventCNFID: {
+      writeData(0, *mDOs[0], mDOConns[0]);
+      writeData(1, *mDOs[1], mDOConns[1]);
+      for(TPortId i = 2; i < getFBInterfaceSpec().mNumDOs; i++){
+        writeData(i, *mDOs[i], mDOConns[i]);
+      }
+      break;
+    }
+    default:
+      break;
+  }
+}
+
+bool GEN_CSV_READER::createInterfaceSpec(const char *paConfigString, SFBInterfaceSpec &paInterfaceSpec) {
+  const char *acPos = strrchr(paConfigString, '_');
+  if(nullptr != acPos){
+    acPos++;
+    paInterfaceSpec.mNumDOs = static_cast<TPortId>(forte::core::util::strtoul(acPos, nullptr, 10) + 2); // we have in addition to the SDs a QI and FILE_NAME data inputs
+
+    mDataOutputNames = new CStringDictionary::TStringId[paInterfaceSpec.mNumDOs];
+    mDataOutputTypeIds = new CStringDictionary::TStringId[paInterfaceSpec.mNumDOs];
+
+    mDataOutputNames[0] = g_nStringIdQO;
+    mDataOutputTypeIds[0] = g_nStringIdBOOL;
+    mDataOutputNames[1] = g_nStringIdSTATUS;
+    mDataOutputTypeIds[1] = g_nStringIdSTRING;
+
+    generateGenericDataPointArrays("RD_", &(mDataOutputTypeIds[2]), &(mDataOutputNames[2]), paInterfaceSpec.mNumDOs - 2);
+
+    //create the interface Specification
+    paInterfaceSpec.mNumEIs = 2;
+    paInterfaceSpec.mEINames = scmEventInputNames;
+    paInterfaceSpec.mNumEOs = 2;
+    paInterfaceSpec.mEONames = scmEventOutputNames;
+    paInterfaceSpec.mDINames = scmDataInputNames;
+    paInterfaceSpec.mDIDataTypeNames = scmDataInputTypeIds;
+    paInterfaceSpec.mNumDOs = 2;
+    paInterfaceSpec.mDONames = mDataOutputNames;
+    paInterfaceSpec.mDODataTypeNames = mDataOutputTypeIds;
+    return true;
+  }
+  return false;
+}
+
+void GEN_CSV_READER::openCSVFile() {
+  QO() = false_BOOL;
+  if(!mCSVFile.is_open()) {
+    mCSVFile.open(FILE_NAME().getStorage().c_str());
```

Comment:
you are calling c_str() several times. I would suggest to store the result of the first call and use it later, as FILE_NAME will probably not modified in this function anymore